### ISSUE TITLE
Components that contain ViewChild need change detection after view init

### ISSUE THREAD

**Issue by gionkunz** [angular/angular#5870]
The following example shows a Component that has a reference to a child directive using a `ViewChild` query. If the input changes and causes `NgFor` to create new components, those will execute all view bindings before the child directive is obtained in the component using the `ViewChild` annotation. After the view child is set and after the `ngAfterViewInit` lifecycle hook was called, there will be no additional change detection as the component thinks he's clean. However, if the component has a reference to a view child and this reference is used within the components view, the component should be dirty after the `ngAfterViewInit` lifecycle hook.

The following example should show all even numbers in blue and all odd numbers in red. Due to the change detection issue, after clicking the "Cause Change" button, there is an inconsistency between the view and the model.

The example can be used as follows to reproduce the issue:
1. Click the "Cause Change" button to change the input array to the `NgFor` directive. `NgFor` will detect the third and fourth element to be new and creates them. The view bindings that access the child directive are execute before the ViewChild was initialized. The component stays checked, even if the ViewChild is referenced within the components view.
2. Click the "Cause Change" button again, where no additional components get created (the list differ does its job), but the view bindings execute again and the issue is gone.

http://plnkr.co/edit/zSiHmyzXcm6Yq1jHJgMn?p=preview
```javascript
import {Component, Directive, Input, ViewChild, NgFor} from 'angular2/angular2';

@Directive({
  selector: '[number]'
})
export class NumberDirective {
  @Input() number;
  
  isEven() {
    return this.number % 2 === 0;
  }
}

@Component({
  selector: 'number-component'
  template: `
    <div [number]="number" 
         [style.color]="getColor()">{{number}}</div>
  `,
  directives: [NumberDirective]
})
export class NumberComponent {
  @ViewChild(NumberDirective) numberDirective;
  @Input() number;
  
  constructor() {
    console.log('Constructed');
  }
  
  ngOnChange() {
    console.log('on change', this.numberDirective);
  }
  
  ngAfterViewInit() {
    console.log('after view init', this.numberDirective);
  }
  
  getColor() {
    console.log('getColor', this.numberDirective);
    return this.numberDirective && this.numberDirective.isEven() ? 'blue' : 'red';
  }
}

@Component({
  selector: 'app'
  template: `
    <number-component *ngFor="#number of numbers" 
            [number]="number"></number-component>
    <button (click)="change()">Cause Change</button>
  `,
  directives: [NumberComponent, NgFor]
})
export class App {
  numbers = [1, 2];
  
  change() {
    this.numbers = [1, 2, 3, 4];
  }
}
```

**Comment by ShiYuanjun-Tim** [comment]
Hi ,maybe you already have the [feat(router): Add directive for active route link,](https://github.com/angular/angular/issues/5877) but i have came across your problem when search something , here i have modify some of your code and maybe this fits your need.

#### two changes
>- [style.color] expression 
>- the ngAfterViewInit method.
    **Here the setTimeout(...) is needed [about why?](https://angular.io/docs/ts/latest/guide/lifecycle-hooks.html#!#afterview), because the change in ngAfterViewInit hook won't reflect in this turn, if i'm wrong,plz correct me ! ** 

```javascript
@Component({
  selector: 'number-component'
  template: `
    <div [number]="number" 
         [style.color]="color">{{number}}</div>
  `,
  directives: [NumberDirective]
})
export class NumberComponent {
....

  ngAfterViewInit() {
    console.log('after view init', this.numberDirective);
    setTimeout(()=>{this.color= this.numberDirective.isEven() ? 'blue' : 'red';},0)
  }
```

**Comment by pkozlowski-opensource** [comment]
@gionkunz wouldn't what you are asking here throw uni-directional data flow out of the window? Do you have real-life example where such mechanism would be needed / useful / required?

**Comment by gionkunz** [comment]
@pkozlowski-opensource Well the issue arised for me when I tried to query for the routerLink directive to get its active state. I agree this is not unidirectional but given that there was no other solution for me to programatically get the routerLink active state, I needed to go down this route.

**Comment by jimitndiaye** [comment]
I'm having this same problem with dynamically created views using ```ViewContainerRef.createComponent```. Setting inputs on the created component results in the model being out of sync with the view which then doesn't update till an event occurs on the inserted view.
I've tried using ViewChildren(selector, {read: ChangeDetectorRef}) to get a reference to the child change detector and mark it for change but that doesn't work.

**Comment by spali** [comment]
Not sure if this is the same as the original requesters meant.
But I would have a working plunk with a real use-case I currently have.

i.e. for a table component, I want a single config object for the consumer of the component.
the table component should update itself on any further changes to this config object properties.
because the consumer of the component can provide templates for rendering the cells, the consumer needs to update this in AfterViewInit hook.
My goal is to not let the consumer think about about component internal stuff. It should just be possible to change any config property at any time and the table component should do the rest.

It's a stripped down example.
It works as long as `var shouldWork: boolean = true;` is in ngAfterViewInit, because it replaces the whole config object. But if I want to handle this for the consumer in the component itself, I don't know a good way to to this.
Just change `var shouldWork: boolean = true;` to `var shouldWork: boolean = false;` and angular complains about the changed expression.

http://plnkr.co/edit/cSJPf400VQm168tzUEO6

Hope this matches the original issue, but couldn't find a good way to handle this without hacks.

**Comment by pkozlowski-opensource** [comment]
@spali it doesn't sound like the original issue. Moreover it is not clear why you are saying this:

> because the consumer of the component can provide templates for rendering the cells, the consumer needs to update this in AfterViewInit hook.

Why do you need to use `AfterViewInit`? To get `TemplateRef`s? If so they are available in the `ngAfterContentInit`. But more importantly your design looks really wired - I would expect people to provide templates as children of your table element.

@jimitndiaye I'm not sure I understand your comment. Please open a new issue with a clear reproduction scenario in a plunker.

I'm going to close this one as this issue is turning into a meta-issue which is about everything and nothing. If you still believe that something needs fixing here please open a new issue with a concrete, real-life use case and a minimal plunker illustrating the issue.

**Comment by spali** [comment]
@pkozlowski-opensource 
Thanks, I will check this with AfterContentInit, just assumed because I used ViewChild, it only available in the AfterViewInit.

I started with the templates as children of the table element, but didn't find a way (probably just don't know) how to mark a template with the column name, to identify it inside the component. I just got a QueryList of TemplateRef's, but have not found any possibility to identify them for the specific columns (columns names are also dynamically). So I can't use a static name for ContentChild in the component. If it would propagate the `#mycolumnname` on the template to the table component, then It would work. But inspecting the object in the debugger didn't let me find anything to identify the Template for which column it is.
